Firebase/InstanceID/FIRInstanceIDKeyPairStore.m

```diff
@@ -365,31 +365,43 @@ - (void)migrateKeyPairCacheIfNeededWithHandler:(void (^)(NSError *error))handler
   self.keyPair = keyPair;
 
   // Either new key pair doesn't exist or it's different than legacy key pair, start the migration.
+
+  // Use a dispatch group to call handler when both private and public keys are updated.
+  dispatch_group_t dispatchGroup = dispatch_group_create();
+  __block NSError *updateKeyRefError;
+
+  dispatch_group_enter(dispatchGroup);
   NSString *privateKeyTag = FIRInstanceIDPrivateTagWithSubtype(kFIRInstanceIDKeyPairSubType);
   [self updateKeyRef:keyPair.publicKey
              withTag:publicKeyTag
              handler:^(NSError *error) {
                if (error) {
                  FIRInstanceIDLoggerError(kFIRInstanceIDMessageCodeKeyPairMigrationError,
                                           @"Unable to migrate key pair from legacy ones.");
+                 updateKeyRefError = error;
                }
-               [self updateKeyRef:keyPair.privateKey
-                          withTag:privateKeyTag
-                          handler:^(NSError *error) {
-                            if (error) {
-                              FIRInstanceIDLoggerError(
-                                  kFIRInstanceIDMessageCodeKeyPairMigrationError,
-                                  @"Unable to migrate key pair from legacy ones.");
-                              return;
-                            }
-                            FIRInstanceIDLoggerDebug(
-                                kFIRInstanceIDMessageCodeKeyPairMigrationSuccess,
-                                @"Successfully migrated the key pair from legacy ones.");
-                            if (handler) {
-                              handler(error);
-                            }
-                          }];
+               dispatch_group_leave(dispatchGroup);
+             }];
+
+  dispatch_group_enter(dispatchGroup);
+  [self updateKeyRef:keyPair.privateKey
```

Comment:
Can we also check the nil in updateKeyRef just in case.

Reply:
I think, it is already checked by `[keyPair isValid]` at line 346. Do you think we need to check it here again?

Reply:
I guess it's fine now that you move the function out of the async class.